### PR TITLE
Refactor get_gas_spent

### LINKED ISSUE #1049: Refactor get_gas_spent to calculate_gas
### File Location(s)

_No response_

### Proposal

`fn calculate_gas(...) -> (gas_spent: u64, required_balance: u64)`

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1049

- rename to calculate_gas_info
- use extrinsics in the function
- remove redundant code
- update rpc calls

@gear-tech/dev 

**Release Notes**: `get_gas_spent` was renamed to `calculate_gas_info` to render its new logic: more details about required gas for execution are now provided by the function. New function argument added that controls whether execution of secondary programs may result in a trap. The function calculates gas amount with greater precision.

### CODE REVIEW

File: pallets/gear/src/lib.rs
Comment (shamilsan) [inline pallets/gear/src/lib.rs:110]: Let's think about names.
- `spent` is the gas limit enough to process the message
- `to_send` is the gas amount to be sent to other actors
- `burnt` is the gas amount that will be deducted from the sender's account for the message processing

So I'd rename `spent` to `min_limit` or similar, what do you think?

File: pallets/gear/src/lib.rs
Comment (shamilsan) [inline pallets/gear/src/lib.rs:522]: How will we distinguish the first call error from the second one?

File: pallets/gear/src/lib.rs
Comment (gshep) [reply pallets/gear/src/lib.rs:110]: Yes, it makes sense

File: pallets/gear/src/lib.rs
Comment (gshep) [reply pallets/gear/src/lib.rs:522]: Frankly speaking I don't think we need to distinguish them: any error means that required gas cannot be determined

**Review by shamilsan — APPROVED** [review]
Awesome work 🔥 
We've discussed returning an error when trying to send gas to the user mailbox but as I see you calc it anyway.

File: pallets/gear/src/lib.rs
Comment (shamilsan) [reply pallets/gear/src/lib.rs:522]: Ok, let's collect feedback from users.

**Comment by shamilsan** [comment]
Also, it is important to merge this PR after the IDEA frontend is ready (let me add @osipov-mit to reviewers).

File: pallets/gear/src/lib.rs
Comment (SabaunT) [inline pallets/gear/src/lib.rs:561]: Problems can arise when testing 2 sequential `submit_program` calls during pallet tests. Discussed

**Review by SabaunT — CHANGES_REQUESTED** [review]
Interim review

File: pallets/gear/src/lib.rs
Comment (breathx) [inline pallets/gear/src/lib.rs:111]:
```suggestion
#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq, scale_info::TypeInfo)]
#[cfg_attr(feature = "std", derive(serde::Deserialize, serde::Serialize))]
pub struct GasInfo {
    pub spent: u64,
    pub to_send: u64,
    pub burned: u64,
}

impl GasInfo {
    pub fn balance_diff(&self) -> u64 {
        self.to_send.saturating_add(self.burned)
    }
}
```

Comment (breathx) [reply pallets/gear/src/lib.rs:111]: I also missed the fact that exactly gas can't be sent to anybody - we can paste here gas_reserved (which will be used for charging rent for anything) and VALUE to be sent, which is u128, so maybe need to consider about renaming GasInfo into some EconomicInfo/BalanceInfo

Comment (breathx) [reply pallets/gear/src/lib.rs:111]: `GasInfo` is lightweight enough for me to be also `Copy`

Comment (breathx) [reply pallets/gear/src/lib.rs:111]: Some docs are also need

File: pallets/gear/src/lib.rs
Comment (breathx) [inline pallets/gear/src/lib.rs:521]: this seems not be used at all rn

Comment (breathx) [reply pallets/gear/src/lib.rs:521]: that's bad workaround for me

File: pallets/gear/src/lib.rs
Comment (breathx) [inline pallets/gear/src/lib.rs:710]: Safety flag for me should indicate do we allow other programs panic

For any case panic in currently checking contract means error and should be so
But for many other cases if our contract handles other contracts panic, this is the correct behaviour.

With safety flag we will or won't fail this fn if panics occurred not in main calculating program

**Review by breathx — CHANGES_REQUESTED** [review]
Yeah, great work!

I just want to clarify moment with return types and append useful feature to rpc call within the PR

This PR also **MUST NOT** be merged, until @osipov-mit approve and create similar changes on JS side, once core team approved it.

**Comment by breathx** [comment]
Release notes to be written also, exactly

File: pallets/gear/src/tests.rs
Comment (gshep) [reply pallets/gear/src/tests.rs:65]: discussed

File: pallets/gear/src/lib.rs
Comment (gshep) [reply pallets/gear/src/lib.rs]: discussed

File: pallets/gear/src/lib.rs
Comment (breathx) [inline pallets/gear/src/lib.rs:112]: ... that we reserve for some other on-chain interactions

File: pallets/gear/src/lib.rs
Comment (breathx) [inline pallets/gear/src/lib.rs:665]: better to keep naming as in other places
```suggestion
            let mut reserved = 0;
```

Comment (breathx) [reply pallets/gear/src/lib.rs:665]: this can be calculated at the end of the processing only once

File: pallets/gear/src/tests.rs
Comment (breathx) [inline pallets/gear/src/tests.rs:813]: Do we really can't avoid it change change the demo a bit?
Or you some proxy contract?

**Review by breathx — APPROVED** [review]
Glad to see such improvements 👍

File: pallets/gear/src/tests.rs
Comment (gshep) [reply pallets/gear/src/tests.rs:813]: Programs in both tests are sending message with value to a user:
`(import "env" "gr_send_wgas" (func $send (param i32 i32 i32 i64 i32 i32) (result i32)))`

and as I suppose it is a required part of test logic.

**Review by osipov-mit — APPROVED** [review]

**Review by SabaunT — APPROVED** [review]

**Review by breathx — APPROVED** [review]
Great

**Pull request merged**
